ext/liquid_c/vm.c

```diff
@@ -11,53 +12,394 @@ ID id_ivar_last_capture_length;
+    "liquid_vm",
+    { vm_mark, vm_free, vm_memsize, },
+    NULL, NULL, RUBY_TYPED_FREE_IMMEDIATELY
+};
+
+static VALUE vm_internal_new(VALUE context)
+{
+    vm_t *vm;
+    VALUE obj = TypedData_Make_Struct(cLiquidCVM, vm_t, &vm_data_type, vm);
+    vm->stack = c_buffer_init();
+
+    vm->strainer = rb_funcall(context, id_strainer, 0);;
+    Check_Type(vm->strainer, T_OBJECT);
+
+    vm->filter_methods = rb_funcall(RBASIC_CLASS(vm->strainer), id_filter_methods_hash, 0);;
+    Check_Type(vm->filter_methods, T_HASH);
+
+    vm->interrupts = rb_ivar_get(context, id_ivar_interrupts);
+    Check_Type(vm->interrupts, T_ARRAY);
+
+    vm->strict_filters = RTEST(rb_funcall(context, id_strict_filters, 0));
+    vm->global_filter = rb_funcall(context, id_global_filter, 0);
+    vm->resource_limits = rb_ivar_get(context, id_ivar_resource_limits);;
+    return obj;
+}
+
+static vm_t *vm_from_context(VALUE context)
+{
+    VALUE vm_obj = rb_attr_get(context, id_vm);
+    if (vm_obj == Qnil) {
+        vm_obj = vm_internal_new(context);
+        rb_ivar_set(context, id_vm, vm_obj);
+    }
+    // instance variable is hidden from C so should be safe to unwrap it without type checking
+    return DATA_PTR(vm_obj);
+}
+
+static void write_fixnum(VALUE output, VALUE fixnum)
+{
+    long long number = RB_NUM2LL(fixnum);
+    int write_length = snprintf(NULL, 0, "%lld", number);
+    long old_size = RSTRING_LEN(output);
+    long new_size = old_size + write_length;
+    long capacity = rb_str_capacity(output);
+
+    if (new_size > capacity) {
+        do {
+            capacity *= 2;
+        } while (new_size > capacity);
+        rb_str_resize(output, capacity);
+    }
+    rb_str_set_len(output, new_size);
+
+    snprintf(RSTRING_PTR(output) + old_size, write_length + 1, "%lld", number);
+}
+
+static void write_obj(VALUE output, VALUE obj)
+{
+    switch (TYPE(obj)) {
+        case T_STRING:
+            rb_str_buf_append(output, obj);
+            break;
+        case T_FIXNUM:
+            write_fixnum(output, obj);
+            break;
+        case T_ARRAY:
+            for (long i = 0; i < RARRAY_LEN(obj); i++)
+            {
+                VALUE item = RARRAY_AREF(obj, i);
+
+                if (RB_UNLIKELY(RB_TYPE_P(item, T_ARRAY))) {
+                    // Normally liquid arrays are flat, but for safety and simplicity we
+                    // leverage ruby's join that detects and raises on a recursion loop
+                    rb_str_buf_append(output, rb_ary_join(item, Qnil));
+                } else {
+                    write_obj(output, item);
+                }
+            }
+            break;
+        case T_NIL:
+            break;
+        default:
+            obj = rb_funcall(obj, id_to_s, 0);
+            rb_str_append(output, obj);
+            break;
     }
+}
+
+static inline void vm_stack_push(vm_t *vm, VALUE value)
+{
+    VALUE *stack_ptr = (VALUE *)vm->stack.data_end;
+    *stack_ptr++ = value;
```

Comment:
We can't really do that and share the stack for nested vm execution (e.g. block rendering or expression evaluation) without setting `vm->stack.data_end` before calling any ruby methods that have access to the context and reading it after returning from those method calls.  Otherwise, those nested vm executions could overwrite values for the parent or resize the stack and invalidate the stack pointer used by the parent.

There are definitely some ways around this problem.  For instance, we could have a separate buffers for nested vm executions, but keep around those buffers for following nested vm execution at the same level of nesting.  Another approach would be to reserve the maximum stack space needed for a vm execution, then have nested vm executions start from the end of that reserved space, but this could also result in additional buffer allocations if we need to avoid invalidating the parent's stack pointer by reallocating the existing buffer.

I ended up deciding on this approach for now, since it seemed to be fairly simple without introducing stack allocation overhead for nested vm execution, which will be more of a concern when using VM execution to evaluate expression objects.

Reply:
> without setting `vm->stack.data_end` before calling any ruby methods that have access to the context and reading it after returning from those method calls.

That's what I meant. Why not just do it before and after ruby method calls? Seems there will be less ruby method calls than stack operations in a VM loop run.